### PR TITLE
TF: XLA logits processors - minimum length, forced eos, and forced bos

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

(Review after https://github.com/huggingface/transformers/pull/16899)

A few more XLA-compatible logits processors -- minimum length, forced eos, and forced bos. Only the first one needed changes, mostly to avoid needless retracing (it actually compiled without changes but would trigger a retrace at iteration, which would be super slow).

After this PR, the only remaining processors are the bad words and ngrams ones.

### CODE REVIEW

File: src/transformers/generation_tf_logits_process.py [hunk 1]
```diff
@@ -215,13 +215,18 @@ def __init__(self, min_length: int, eos_token_id: int):
...
         self.eos_token_id = eos_token_id
 
-    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.Tensor:
-        # TODO(Matt) - this if statement has to be rewritten for XLA. Leaving it now though since
-        # generate is not XLA - compileable anyways
-        if cur_len < self.min_length:
-            eos_token_id_mask = tf.broadcast_to(tf.range(scores.shape[-1]) == self.eos_token_id, scores.shape)
-            scores = tf.where(eos_token_id_mask, float("-inf"), scores)
+    def _apply_eos_token_mask(self, scores: tf.Tensor) -> tf.Tensor:
+        eos_token_id_mask = tf.broadcast_to(tf.range(scores.shape[-1]) == self.eos_token_id, scores.shape)
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_logits_process.py:219]:
```suggestion
        eos_token_id_mask = tf.range(scores.shape[-1]) == self.eos_token_id
```
The array should broadcast to the shape of `scores` in the next line anyway without an explicit `broadcast_to`!

File: src/transformers/generation_tf_logits_process.py [hunk 2]
```diff
@@ -215,13 +215,18 @@ def __init__(self, min_length: int, eos_token_id: int):
...
+        scores = tf.where(eos_token_id_mask, float("-inf"), scores)
+        return scores
 
+    def __call__(self, input_ids: tf.Tensor, scores: tf.Tensor, cur_len: int) -> tf.Tensor:
+        # applies eos token masking if the first argument is true
+        scores = tf.cond(
+            tf.less(cur_len, self.min_length),
+            lambda: self._apply_eos_token_mask(scores),
+            lambda: tf.identity(scores),
```
Comment (Rocketknight1) [inline src/transformers/generation_tf_logits_process.py:228]:
```suggestion
            self._apply_eos_token_mask(scores),
            scores,
```
Would this work without the lambdas and identity call? I feel like it should but I'm not sure if I'm missing something obvious.

**Review by Rocketknight1 — APPROVED** [review]
LGTM! I have a couple of nits that shouldn't actually affect the output, and one of them might be me being totally wrong anyway.

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me!

File: src/transformers/generation_tf_logits_process.py (hunk 2 above)
Comment (gante) [reply src/transformers/generation_tf_logits_process.py:228]: Nope, it doesn't. Super unintuitive, but `tf.cond` expects a callable, not the output of each branch 😬 ([docs](https://www.tensorflow.org/api_docs/python/tf/cond))

It fails if we remove the lambda.
